microprofile/config-smallrye/src/main/java/org/wildfly/extension/microprofile/config/smallrye/deployment/DependencyProcessor.java

```diff
@@ -0,0 +1,78 @@
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+
+package org.wildfly.extension.microprofile.config.smallrye.deployment;
+
+import org.jboss.as.server.deployment.Attachments;
+import org.jboss.as.server.deployment.DeploymentPhaseContext;
+import org.jboss.as.server.deployment.DeploymentUnit;
+import org.jboss.as.server.deployment.DeploymentUnitProcessor;
+import org.jboss.as.server.deployment.Phase;
+import org.jboss.as.server.deployment.module.ModuleDependency;
+import org.jboss.as.server.deployment.module.ModuleSpecification;
+import org.jboss.logging.Logger;
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleIdentifier;
+import org.jboss.modules.ModuleLoader;
+
+/**
+ * Add dependencies required by deployment unit to access the Config API (programmatically or using CDI).
+ */
+public class DependencyProcessor implements DeploymentUnitProcessor {
+
+    Logger log = Logger.getLogger(SubsystemDeploymentProcessor.class);
+
+    /**
+     * See {@link Phase} for a description of the different phases
+     */
+    public static final Phase PHASE = Phase.DEPENDENCIES;
+
+    /**
+     * The relative order of this processor within the {@link #PHASE}.
+     * The current number is large enough for it to happen after all
+     * the standard deployment unit processors that come with JBoss AS.
+     */
+    public static final int PRIORITY = 0x4000;
+
+    public static final ModuleIdentifier MICROPROFILE_CONFIG_API = ModuleIdentifier.create("org.eclipse.microprofile.config.api");
+    public static final ModuleIdentifier MICROPROFILE_CONFIG_EXTENSION = ModuleIdentifier.create("org.wildfly.extension.microprofile.config-smallrye");
+
+    @Override
+    public void deploy(DeploymentPhaseContext phaseContext) {
+        DeploymentUnit deploymentUnit = phaseContext.getDeploymentUnit();
+
+        addDependencies(deploymentUnit);
+    }
+
+    @Override
+    public void undeploy(DeploymentUnit context) {
+    }
+
+    private void addDependencies(DeploymentUnit deploymentUnit) {
+        final ModuleSpecification moduleSpecification = deploymentUnit.getAttachment(Attachments.MODULE_SPECIFICATION);
+        final ModuleLoader moduleLoader = Module.getBootModuleLoader();
+
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_API, false, false, true, false));
+        moduleSpecification.addSystemDependency(new ModuleDependency(moduleLoader, MICROPROFILE_CONFIG_EXTENSION, false, false, true, false));
```

Comment:
Exposing a subsystem module to a deployment isn't good. We should get clean separation between deployment code and management code, e.g. so we can do things like https://docs.google.com/document/d/1OK8izSS9DVSOAqXjadG7piIjWOJ4KU2eKQ548ZL26aM/edit#heading=h.n28b42dwjj5x

Reply:
fixed. the deployments no longer has any dependency on the extension module.